File: guests/swift/richtext.swift
// The rich text scene, Swift port — guests/rust/richtext.rs,
// tools/scenes/richtext.steps. THE OFFSETS ARE UTF-8 BYTES: the é in the
// first word is what makes a UTF-16 reader fail (docs/ranges-units.md).

import Foundation

let document = "Héllo world\nSecond line"

/// The core's spelling of runs (`expect_runs`), so the binding's document
/// and the core's mirror are compared as one string.
func spell(_ runs: [KayaRun]) -> String {
    runs.map { run in
        run.value == "true"
            ? "\(run.start):\(run.end) \(run.name)"
            : "\(run.start):\(run.end) \(run.name)=\(run.value)"
    }.joined(separator: "|")
}

let app = KayaApp()

app.build { tx in
    tx.window(title: "richtext")
    let last = tx.signal(.str(""))
    let runs = tx.signal(.str(""))

    // A widget parents at CREATION, so the editor rides out through this
    // var (docs/traps.md, result builders).
    var editor: KayaWidget! = nil

    let root = tx.column {
        editor = tx.textarea(
            rich: true,
            onEdit: { t, edit in
                let mirror = spell(app.document(editor).runs)
                t.write(
                    last,
                    .str("edit \(edit.start):\(edit.end) <\(edit.inserted)> "
                        + "[\(spell(edit.runs))]"))
                t.write(runs, .str(mirror))
            },
            onFormat: { t, act in
                let mirror = spell(app.document(editor).runs)
                t.write(
                    last,
                    .str("format \(act.start):\(act.end) \(act.name)="
                        + "\(act.value ?? "off")"))
                t.write(runs, .str(mirror))
            })
        tx.setA11yId(editor, "doc")
        tx.setA11yLabel(editor, "Document")
        tx.label(bind: last)  // label#0
        tx.label(bind: runs)  // label#1
        tx.row {
            tx.button("seed") { t in  // button#0
                let doc = KayaDocument(document)
                    .bold(0..<6)
                    .link(7..<12, "https://kaya.dev")
                    .block(13..<24, .heading2)
                t.setDocument(editor, doc)
                t.write(runs, .str(spell(doc.runs)))
            }
            tx.button("insert") { t in  // button#1
                let edit = KayaEdit.insert(at: 6, ", big").mark(2..<5, "italic", "true")
                t.applyEdit(editor, edit)
                t.write(runs, .str(spell(app.document(editor).runs)))
            }
            tx.button("select word") { t in  // button#2
                t.selectRange(editor, 0..<6)
            }
            tx.button("unbold") { t in  // button#3
                t.unformat(editor, "bold")
            }
            tx.button("heading") { t in  // button#4
                t.setBlock(editor, .heading1)
            }
            tx.button("focus") { t in  // button#5
                t.focus(editor)
            }
            tx.button("prefix") { t in  // button#6
                t.applyEdit(editor, KayaEdit.insert(at: 0, "> "))
                t.write(runs, .str(spell(app.document(editor).runs)))
            }
        }
    }
    tx.mount(root)
}

app.run()

File: guests/ocaml/richtext.ml
(* The rich text scene, OCaml port — guests/rust/richtext.rs,
   tools/scenes/richtext.steps. THE OFFSETS ARE UTF-8 BYTES; the é in the
   first word is what makes a UTF-16 reader fail
   (docs/ranges-units.md). *)

open Kaya_app

let doc_source = "Héllo world\nSecond line"

(* The core's spelling of runs ([expect_runs]), so the binding's document
   and the core's mirror are compared as one string. *)
let spell (runs : Run.t list) =
  String.concat "|"
    (List.map
       (fun (r : Run.t) ->
         if r.value = "true" then
           Printf.sprintf "%d:%d %s" r.start r.stop r.name
         else Printf.sprintf "%d:%d %s=%s" r.start r.stop r.name r.value)
       runs)

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
      window ~title:"richtext" ();
      let last = signal_str ("") in
      let runs = signal_str ("") in

      let editor =
        textarea ~rich:true ~a11y_id:"doc" ~a11y_label:"Document" ()
      in
      (* The two handlers read the editor's own folded document, so they
         are registered once it exists. *)
      on_edit app editor (fun e ->
          let mirror = spell (document editor).runs in
          write last
            (Printf.sprintf "edit %d:%d <%s> %s [%s]" e.start e.stop
               e.inserted
               (match e.source with
               | Some s -> edit_source_name s
               | None -> "?")
               (spell e.runs));
          write runs mirror);
      on_format app editor (fun act ->
          let mirror = spell (document editor).runs in
          write last
            (Printf.sprintf "format %d:%d %s=%s" act.start act.stop
               act.name
               (Option.value act.value ~default:"off"));
          write runs mirror);

      mount
        (column
           [
             w editor (* textarea#0 *);
             label ~bind:last (* label#0 *);
             label ~bind:runs (* label#1 *);
             row
               [
                 (* button#0 — the declaration: text and runs in one write *)
                 button ~text:"seed"
                   ~on_click:(fun () ->
                     let doc =
                       Document.create doc_source
                       |> Document.bold (0, 6)
                       |> Document.link (7, 12) "https://kaya.dev"
                       |> Document.block (13, 24) Heading2
                     in
                     set_document editor doc;
                     write runs ((spell doc.runs)));
                 (* button#1 — the app's own edit, italic over the
                    inserted word *)
                 button ~text:"insert"
                   ~on_click:(fun () ->
                     apply_edit editor
                       (Edit.insert 6 ", big" |> Edit.mark (2, 5) "italic" "true");
                     write runs ((spell (document editor).runs)));
                 (* button#2 — select the first word, for the toolbar act *)
                 button ~text:"select word"
                   ~on_click:(fun () -> select_range editor (0, 6));
                 (* button#3 — the toolbar: take bold off the selection *)
                 button ~text:"unbold"
                   ~on_click:(fun () -> unformat editor "bold");
                 (* button#4 — a block act over the selection's paragraphs *)
                 button ~text:"heading"
                   ~on_click:(fun () -> set_block editor Heading1);
                 (* button#5 — focus, so the next keystroke is a USER edit *)
                 button ~text:"focus" ~on_click:(fun () -> focus editor);
                 (* button#6 — clicked while a composition is live: the
                    core holds the edit, the app's document takes it now *)
                 button ~text:"prefix"
                   ~on_click:(fun () ->
                     apply_edit editor (Edit.insert 0 "> ");
                     write runs ((spell (document editor).runs)));
               ];
           ]
           ()));

  exit (run app)
